hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvMetadataAvroResolver.java

```diff
@@ -33,21 +36,54 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.Set;
 import java.util.stream.Stream;
 
 import static com.hazelcast.jet.sql.impl.connector.SqlConnector.AVRO_FORMAT;
 import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_KEY_AVRO_RECORD_NAME;
+import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_KEY_AVRO_SCHEMA;
 import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_VALUE_AVRO_RECORD_NAME;
+import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_VALUE_AVRO_SCHEMA;
+import static com.hazelcast.jet.sql.impl.connector.file.AvroResolver.unwrapNullableType;
 import static com.hazelcast.jet.sql.impl.connector.keyvalue.KvMetadataResolver.extractFields;
 import static com.hazelcast.jet.sql.impl.connector.keyvalue.KvMetadataResolver.maybeAddDefaultField;
+import static com.hazelcast.jet.sql.impl.inject.AvroUpsertTarget.CONVERSION_PREFS;
+import static com.hazelcast.sql.impl.type.converter.Converters.getConverter;
+import static java.util.stream.Collectors.toSet;
 
 public final class KvMetadataAvroResolver implements KvMetadataResolver {
 
     public static final KvMetadataAvroResolver INSTANCE = new KvMetadataAvroResolver();
 
-    private KvMetadataAvroResolver() {
+    /**
+     * Avro is an optional dependency for SQL, so the schemas should be the initialized lazily.
+     */
+    public static class Schemas {
+        /**
+         * {@link QueryDataType#OBJECT} is mapped to the union of all primitive types and null.
+         * Avro {@link Schema.Type#RECORD}s can only be represented using user-defined types.
+         */
+        public static final Schema OBJECT_SCHEMA = SchemaBuilder.unionOf()
+                .nullType()
+                .and().booleanType()
+                .and().intType()
+                .and().longType()
+                .and().floatType()
+                .and().doubleType()
+                .and().stringType()
+                .endUnion();
+
+        private static final Map<QueryDataTypeFamily, List<Schema.Type>> CONVERSIONS =
+                CONVERSION_PREFS.entrySet().stream()
+                        .collect(ImmutableMap::<QueryDataTypeFamily, List<Schema.Type>>builder,
```

Comment:
you can do grouping in streams easier:
```
                        .collect(Collectors.groupingBy(e -> getConverter(e.getKey()).getTypeFamily(),
                                        Collectors.flatMapping(e -> e.getValue().stream(), toList())));
```
you would have to add `UNION` to that.

Reply:
or possibly even simpler:
```java
Stream.concat(
                        CONVERSION_PREFS.entrySet().stream()
                                .map(e -> Map.entry(getConverter(e.getKey()).getTypeFamily(),e.getValue())),
                        Stream.of(Map.entry(QueryDataTypeFamily.OBJECT, List.of(Schema.Type.UNION, Schema.Type.NULL)))
                    )
                    .collect(toUnmodifiableMap(Entry::getKey, Entry::getValue));
```

Reply:
I am mapping entries without doing any grouping, so the first alternative is confusing. I don't understand in what way the second alternative is "simpler."

Reply:
simpler because you do not have to use Map builder. But that is mostly matter of taste, so we can keep the code as is.